packages/eslint-plugin/src/rules/thenable-in-promise-aggregators.ts

```diff
@@ -0,0 +1,127 @@
+  meta: {
+    docs: {
+      description:
+        'Disallow passing non-Thenable values to promise aggregators',
+      recommended: 'recommended',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      inArray:
+        'Unexpected non-Thenable value in array passed to promise aggregator.',
+      arrayArg:
+        'Unexpected array of non-Thenable values passed to promise aggregator.',
+      nonArrayArg: 'Unexpected non-array passed to promise aggregator.',
+    },
+    schema: [],
+    type: 'problem',
+  },
+  defaultOptions: [],
+
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+
+    return {
+      CallExpression(node: TSESTree.CallExpression): void {
+        if (node.callee.type !== AST_NODE_TYPES.MemberExpression) {
+          return;
+        }
+        if (node.callee.object.type !== AST_NODE_TYPES.Identifier) {
+          return;
+        }
+        if (node.callee.object.name !== 'Promise') {
+          return;
+        }
+        if (node.callee.property.type !== AST_NODE_TYPES.Identifier) {
+          return;
+        }
+
+        const { name } = node.callee.property;
+        if (!['race', 'all', 'allSettled'].includes(name)) {
+          return;
+        }
+
+        const { arguments: args } = node;
+        if (args.length !== 1) {
+          return;
+        }
+
+        const arg = args[0];
+        if (arg.type === AST_NODE_TYPES.ArrayExpression) {
+          const { elements } = arg;
+          if (elements.length === 0) {
+            return;
+          }
+
+          for (const element of elements) {
+            if (element == null) {
+              continue;
+            }
+            const elementType = services.getTypeAtLocation(element);
+            if (isTypeAnyType(elementType) || isTypeUnknownType(elementType)) {
+              continue;
+            }
+
+            const originalNode = services.esTreeNodeToTSNodeMap.get(element);
+            if (tsutils.isThenableType(checker, originalNode, elementType)) {
+              continue;
+            }
+
+            context.report({
+              messageId: 'inArray',
+              node: element,
+            });
+          }
+        } else {
+          const argType = services.getTypeAtLocation(arg);
+          if (isTypeAnyType(argType) || isTypeUnknownType(argType)) {
+            return;
+          }
+
+          if (!checker.isArrayType(argType)) {
+            context.report({
+              messageId: 'nonArrayArg',
+              node: arg,
+            });
+            return;
+          }
+
+          if (argType.typeArguments === undefined) {
+            return;
+          }
+
+          if (argType.typeArguments.length < 1) {
+            return;
+          }
+
+          const typeArg = argType.typeArguments[0];
+          if (isTypeAnyType(typeArg) || isTypeUnknownType(typeArg)) {
+            return;
+          }
```

Comment:
It'd be cool if we had the following test cases to fully test this condition
```ts
declare const arrOfAny: any[]
Promise.all(arrOfAny)
```
```ts
declare const arrOfUnknown: unknown[]
Promise.all(arrOfUnknown)
```

Reply:
Added the tests. I'm noticing that you used the syntax I found for the invalid ones for a valid case; can/should I switch the valid tests from the current

```ts
async function test() {
  // ...
}
```

to this style, like in the invalid cases?

Reply:
It's not mandatory, but, personally, I prefer to avoid extra code in test cases. 
For example: there are two identical test cases, but the second one takes 5x more lines than the first one:

```ts
ruleTester.run('thenable-in-promise-aggregators', rule, {
  valid: [
    `await Promise.race([Promise.resolve(3)]);`,
    `
async function test() {
  await Promise.race([Promise.resolve(3)]);
}
    `,
  ]
})
```

(for multiline arranges + assertions it only adds two extra lines, but whatever)

We can wait for the maintainer to make a final decision, but in general I think it's fine to have cases with
```ts
async function test() {
  // ...
}
```

Reply:
Alright great! Wasn't sure if there was a particular reason for the function in valid cases; I'm updating to the shortened version